File: app/src/main/java/com/itachi1706/busarrivalsg/RecyclerViews/BusStopRecyclerAdapter.java
package com.itachi1706.busarrivalsg.RecyclerViews;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.analytics.FirebaseAnalytics;
import com.itachi1706.busarrivalsg.BusServicesAtStopRecyclerActivity;
import com.itachi1706.busarrivalsg.R;
import com.itachi1706.busarrivalsg.gsonObjects.sgLTA.BusStopJSON;
import com.itachi1706.busarrivalsg.util.ShortcutHelper;
import com.itachi1706.helperlib.helpers.LogHelper;

import java.util.List;
import java.util.Locale;

/**
 * Created by Kenneth on 31/10/2015.
 * for SingBuses in package com.itachi1706.busarrivalsg.RecyclerViews
 */
public class BusStopRecyclerAdapter extends RecyclerView.Adapter<BusStopRecyclerAdapter.BusStopViewHolder> {

    /**
     * This recycler adapter is used in the internal retrive all bus services from bus stop activity
     */

    private List<BusStopJSON> items;

    public BusStopRecyclerAdapter(List<BusStopJSON> objectList){
        this.items = objectList;
    }

    public void updateAdapter(List<BusStopJSON> newObjects){
        this.items = newObjects;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public BusStopViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View busServiceView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.recyclerview_bus_stops, parent, false);
        return new BusStopViewHolder(busServiceView);
    }

    @Override
    public void onBindViewHolder(BusStopViewHolder holder, int position) {
        BusStopJSON i = items.get(position);

        holder.stopName.setText(i.getDescription());
        holder.desc.setText((!i.isHasDistance()) ? i.getRoadName() + " (" + i.getBusStopCode() + ")" : String.format(Locale.getDefault(), "%s (%s) [%.2fm]",i.getRoadName(), i.getBusStopCode(), i.getDistance()));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void handleClick(Context context, BusStopJSON stop) {
        LogHelper.d("Size", "" + items.size());
        Intent serviceIntent = new Intent(context, BusServicesAtStopRecyclerActivity.class);
        serviceIntent.putExtra("stopCode", stop.getBusStopCode());
        serviceIntent.putExtra("stopName", stop.getDescription());
        serviceIntent.putExtra("busServices", stop.getServices());
        context.startActivity(serviceIntent);

        FirebaseAnalytics mFirebaseAnalytics = FirebaseAnalytics.getInstance(context);
        Bundle bundle = new Bundle();
        bundle.putString(FirebaseAnalytics.Param.ITEM_ID, "Code: " + stop.getBusStopCode());
        bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, "openBusStopDetail");
        mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle);

        // Add dynamic shortcuts
        ShortcutHelper shortcutHelper = new ShortcutHelper(context);
        shortcutHelper.updateBusStopShortcuts(stop, serviceIntent);
    }


    class BusStopViewHolder extends RecyclerView.ViewHolder implements OnClickListener {

        TextView stopName;
        TextView desc;

        BusStopViewHolder(View v){
            super(v);
            stopName = v.findViewById(R.id.tvName);
            desc = v.findViewById(R.id.tvSubText);
            v.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            int position = this.getLayoutPosition();
            final BusStopJSON item = items.get(position);
            handleClick(v.getContext(), item);
        }
    }
}

File: app/src/main/java/com/itachi1706/busarrivalsg/adapters/BusStopRecyclerAdapter.kt
package com.itachi1706.busarrivalsg.adapters

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.analytics.FirebaseAnalytics
import com.itachi1706.busarrivalsg.BusServicesAtStopRecyclerActivity
import com.itachi1706.busarrivalsg.databinding.RecyclerviewBusStopsBinding
import com.itachi1706.busarrivalsg.objects.json.ltasg.BusStopJSON
import com.itachi1706.busarrivalsg.util.ShortcutHelper
import com.itachi1706.helperlib.helpers.LogHelper.d
import java.util.Locale

/**
 * This recycler adapter is used in the internal retrieval all bus services from bus stop activity
 */
class BusStopRecyclerAdapter(private var items: MutableList<BusStopJSON>) :
    RecyclerView.Adapter<BusStopRecyclerAdapter.BusStopViewHolder>() {

    fun updateAdapter(newObjects: List<BusStopJSON>?) {
        val oldSize = items.size
        items = newObjects?.toMutableList() ?: mutableListOf()
        if (oldSize == items.size) {
            notifyItemRangeChanged(0, items.size)
        } else {
            notifyItemRangeRemoved(0, oldSize)
            notifyItemRangeInserted(0, items.size)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BusStopViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = RecyclerviewBusStopsBinding.inflate(inflater, parent, false)
        return BusStopViewHolder(binding)
    }

    override fun onBindViewHolder(holder: BusStopViewHolder, position: Int) {
        val i = items[position]

        holder.binding.tvName.text = i.description
        holder.binding.tvSubText.text =
            if (!i.isHasDistance) "${i.roadName} (${i.busStopCode})" else String.format(
                Locale.getDefault(), "%s (%s) [%.2fm]", i.roadName, i.busStopCode, i.distance
            )
    }

    override fun getItemCount(): Int {
        return items.size
    }

    fun handleClick(context: Context?, stop: BusStopJSON) {
        d("Size", "${items.size}")
        val serviceIntent = Intent(context, BusServicesAtStopRecyclerActivity::class.java).apply {
            putExtra("stopCode", stop.busStopCode)
            putExtra("stopName", stop.description)
            putExtra("busServices", stop.services)
        }
        context?.startActivity(serviceIntent)

        if (context == null) {
            return // No more operation
        }

        val mFirebaseAnalytics = FirebaseAnalytics.getInstance(context)
        val bundle = Bundle().apply {
            putString(FirebaseAnalytics.Param.ITEM_ID, "Code: ${stop.busStopCode}")
            putString(FirebaseAnalytics.Param.CONTENT_TYPE, "openBusStopDetail")
        }
        mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle)

        // Add dynamic shortcuts
        val shortcutHelper = ShortcutHelper(context)
        shortcutHelper.updateBusStopShortcuts(stop, serviceIntent)
    }

    inner class BusStopViewHolder(val binding: RecyclerviewBusStopsBinding) :
        RecyclerView.ViewHolder(binding.root), View.OnClickListener {

        init {
            binding.root.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            val position = layoutPosition
            val item = items[position]
            handleClick(v.context, item)
        }
    }
}